components/authz-service/engine/opa/opa.go

```diff
@@ -621,19 +621,45 @@ func (s *State) SetPolicies(ctx context.Context, policies map[string]interface{}
 	return s.initPartialResult(ctx)
 }
 
-// Spike TODO: Can we have a separate method for just setting the rule mappings?
-// OR does the entire OPA store have to be re-evaluated at once. IF that's true,
-// should we have the same OPA instance in general for rules?
+// Q: Can we have a separate method for just setting the rule mappings?
+//    OR does the entire OPA store have to be re-evaluated at once. IF that's
+//    true, should we have the same OPA instance in general for rules?
+// A: This depends on whether it's used: If we imagine the rules-related queries
+//    to eventually check policies and their projects, it needs to be in the
+//    same store.
 //
-// V2SetPolicies replaces OPA's data with a new set of policies and roles,
-// and resets the partial evaluation cache for v2
+// V2SetPolicies replaces OPA's data with a new set of policies, rules, and
+// roles, and resets the partial evaluation cache for v2
 func (s *State) V2SetPolicies(
 	ctx context.Context, policyMap map[string]interface{},
 	roleMap map[string]interface{}, ruleMap map[string][]interface{}) error {
+
+	// here, we flatten the policy structure: let's see if it buys us something in
+	// opa performance
+	stmts := map[string]interface{}{}
+	for _, pol := range policyMap {
+		pol, ok := pol.(map[string]interface{})
+		if !ok {
+			return errors.New("unexpected policyMap")
+		}
+		for sid, stmt := range pol["statements"].(map[string]interface{}) {
+			if _, ok := stmts[sid]; ok {
+				return errors.New("statement id clash")
+			}
+			stmt := stmt.(map[string]interface{})
+			stmt["type"] = pol["type"]
+			stmt["members"] = pol["members"]
+			stmts[sid] = stmt
+
+			// Note: these aren't currently made use of, so we don't copy them into the
+			// statements
+			// stmt["pol_id"] = id // policy id, key of policyMap above
```

Comment:
This lacks the context from your preamble. Consider saying something to the effect, quoting you, "should we ever need the policy ID, for example to return it to indicate why something was denied, it could also be passed on to the statements"